#7930 part2 add open folder button

Osmodium opened the pull request:
This is "part 2" of #7960 PR.

Added a button to the toolbar for easy access to the custom user content folder.

Currently only win32.

Osmodium commented:
Did a "rebase" :)

Gymnasiast commented:
Ok, macOS and Linux support will need to be added before this get merged. Other people can help with this.

In the case of Linux/BSD, I'd recommend using `xdg-open` to open the folder - it ensures the correct file manager is used.

Osmodium commented:
I will have a look at macOS, but I do need help with the Linux part.

Gymnasiast commented on src/openrct2-ui/windows/TitleMenu.cpp:
```diff
@@ -202,6 +203,9 @@ static void window_title_menu_dropdown(rct_window* w, rct_widgetindex widgetInde
                 Editor::LoadTrackManager();
                 break;
+            case 4:
+                context_open_custom_user_content_folder();
```
For consistency, please also rename the functions accordingly (you already renamed the ones in the previous PR at my request).

Gymnasiast commented on src/openrct2-ui/windows/TitleMenu.cpp:
```diff
@@ -204,7 +204,7 @@ static void window_title_menu_dropdown(rct_window* w, rct_widgetindex widgetInde
                 break;
             case 4:
-                context_open_custom_user_content_folder();
+                context_open_custom_content_folder();
```
Hm, I see you followed a remark that I posted, but I removed that within a minute, since I wasn't sure. Sorry about that.

I'll look into it myself, since I'll have to add Linux support anyway.

Osmodium replied:
Oh, I got an email about it, so I just thought I was blind haha.

Osmodium replied:
Can just revert the commit if necessary :)

Gymnasiast replied:
Nah, I'll do that. It's not worth refiring the CI just to remove the commit - I'll do it when I do the definitive rename and add Linux support.

Osmodium commented:
To clarify why the mehod called is 'context_open_custom_content_folder' and is part of context.h is due to the fact that TitleMenu.cpp doesn't reference platform.h (where this method would belong in my opinion). So instead of including platform.h (and everything in it), I went ahead and added it to the context instead.
Also I added the method 'platform_open_folder' to platform.h so there is a general method for this at a later point, though I can remove that, since it isn't used at the moment.

Gymnasiast commented:
> To clarify why the mehod called is context_open_custom_content_folder and is part of context.h is due to the fact that TitleMenu.cpp doesn't include platform.h (where this method would belong in my opinion). So instead of including platform.h (and everything in it), I went ahead and added it to the context instead.
Also I added the method platform_open_folder to platform.h so there is a general method for this at a later point, though I can remove that, since it isn't used at the moment.

@IntelOrca Could you say if this is the correct approach? I have a few doubts but I'm not sure enough to say so.

IntelOrca commented:
* You don't need a wrapper in Context, just call the UiContext method directly.
* There is no further need to write `context_` like methods as whole code base is C++.
* Use `ShellExecuteW` to support unicode paths.

IntelOrca commented:
@Osmodium Just `GetContext()->GetUiContext()` should work. Make sure you have Context.h and UiContext.h included and that you are using namespace OpenRCT2.

Osmodium commented:
@IntelOrca In order to make it work, I had to include both `openrct2/ui/UiContext.h`
`openrct2/PlatformEnvironment.h` and the implementation looks to me like something that should be somewhere else. But I guess if this is how you want it :)
![image](https://user-images.githubusercontent.com/234248/45041449-a2d8ba80-b068-11e8-9737-0f72436610be.png)

IntelOrca commented:
@Osmodium Yea, your screenshot is how I would have done it.

Osmodium commented:
Okay, so it doesn't matter that we include 2 extra header files in the `TitleMenu.cpp` "just" to avoid having a method in Context.cpp that already has them included? Wouldn't this produce some sort of overhead, or am I overthinking it? :)

Gymnasiast commented:
I have added Linux support. I also rebased your branch. A thing to note: never merge a branch into yours when doing a PR, always rebase.

Osmodium commented:
@Gymnasiast Great and thanks! I haven't done so much rebasing, but I will do this in the future.

Gymnasiast commented on src/openrct2-ui/UiContext.Linux.cpp:
```diff
@@ -113,6 +113,12 @@ namespace OpenRCT2::Ui
 
+        void OpenFolder(const std::string& path) override
+        {
+            std::string cmd = String::Format("xdg-open \"%s\"", path.c_str());
```
It's already between double quotes, which is how we handle most cases. (And Zenity/dialog doesn't feature escaping at all!)

But if you think this escaping is insufficient, could you give more directions on how to do it, or add it yourself?

IntelOrca commented on src/openrct2-ui/UiContext.Linux.cpp:
```diff
@@ -376,6 +376,29 @@ namespace OpenRCT2::Ui
+            *outputPtr = '"';
+
+            return std::string(output);
+        }
```
If path is longer than 256 then you are going to exceed the fixed output buffer. I suggest creating a `std::string` of capacity 256 and appending to that.

AaronVanGeffen replied:
Since you're passing a std::string by reference anyway, there's no need to do iterate the buffer the C way. Instead, you can do something like:

```
size_t index = 0;
while ((index = path.find('"', index) != std::string::npos)
{
     path.replace(index, index + 1, 2, "\\\"");
     index += 2;
}
```

Osmodium commented:
Is there still an issue with the Linux way of getting the correct escaped path, or is is just the changelog that needs to be changed?

Gymnasiast commented:
The Linux bit still has to be changed, but I don't quite know how to follow the instructions that @IntelOrca and @AaronVanGeffen gave.

AaronVanGeffen commented:
I've applied my suggestion to simplify the EscapePathForShell implementation. While testing, I noticed the following:

![screenshot_20180913_222917](https://user-images.githubusercontent.com/604665/45514160-ec31b400-b7a4-11e8-9d4a-7203c0415f95.png)

The toolbox menu is now partially off-screen. We should probably move the buttons up a little. Thoughts?

Osmodium commented:
Beautiful! I don't know how I did not notice that it was a bit off-screen, but thanks for the adjustment :)

janisozaur commented:
This broke Android builds. Please fix it.

Gymnasiast commented:
Android build barely work if they build. Also, I'm sure you would be best person to do that job - you know most about it (along with Marijn).

IntelOrca commented:
It is probably just a case of creating a stub for OpenFolder in the Android UI context.
